cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/RemoteProgramRunDispatcher.java

```diff
@@ -113,20 +111,30 @@ public ProgramController dispatchProgram(ProgramRunDispatcherInfo programRunDisp
         ? programRunnerFactory
         : Optional.ofNullable(remoteProgramRunnerFactory).orElseThrow(UnsupportedOperationException::new)).create(
         programId.getType());
+    if (!(runner instanceof ProgramControllerCreator)) {
+      String msg = String.format("Program %s with runid %s uses an unsupported controller for remote dispatching.",
+                                 programRunDispatcherInfo.getProgramDescriptor().getProgramId(),
+                                 programRunDispatcherInfo.getRunId());
+      throw new Exception(msg);
```

Comment:
We should avoid `throw new Exception()` when possible, UnsupportedOperationException is a little more specific